tls/s2n_record.h

```diff
@@ -16,11 +16,55 @@
 #pragma once
 
 #include <stdint.h>
+#include "crypto/s2n_hmac.h"
+#include "stuffer/s2n_stuffer.h"
 
-#include "s2n_connection.h"
+#define S2N_TLS_CONTENT_TYPE_LENGTH             1
 
-#define TLS13_CONTENT_TYPE_LENGTH 1
+/* All versions of TLS define the record header the same:
+ * ContentType + ProtocolVersion + length
+ */
+#define S2N_TLS_RECORD_HEADER_LENGTH            (S2N_TLS_CONTENT_TYPE_LENGTH + S2N_TLS_PROTOCOL_VERSION_LEN + 2)
+
+/*
+ * All versions of TLS limit the data fragment to 2^14 bytes.
+ *
+ *= https://tools.ietf.org/rfc/rfc5246#section-6.2.1
+ *# The record layer fragments information blocks into TLSPlaintext
+ *# records carrying data in chunks of 2^14 bytes or less.
+ *
+ *= https://tools.ietf.org/rfc/rfc8446#section-5
+ *# The record layer fragments information blocks into TLSPlaintext
+ *# records carrying data in chunks of 2^14 bytes or less.
+ */
+#define S2N_TLS_MAXIMUM_FRAGMENT_LENGTH         (1 << 14)
+
+/* TLS record length allows for 2048 bytes of compression expansion and padding.
+ * However, S2N does not support compression, so we can ignore the compression overhead.
+ */
+#define S2N_TLS12_ENCRYPTION_OVERHEAD_SIZE      1024
```

Comment:
nit: am curious what is driving the need for 1024 here if we've removed compression requirements. worth a follow up doc update.

Reply:
It's the maximum encryption overhead. In TLS1.3 they dropped it to 255, but I guess TLS1.2 was being more conservative. We could probably consider dropping it more since I don't think any of the TLS1.2 cipher actually produce anywhere near 1K overhead, but that would be deviating from the RFC and probably require some research.

Reply:
totally. happy to discuss it as a followup. If the real max for > 1.3 is closer to 255 than 1024, there's further gains here. On both in and out stuffers.